lib/publish-rsr.js

```diff
@@ -0,0 +1,96 @@
+import * as Sentry from '@sentry/node'
+import timers from 'node:timers/promises'
+import pRetry from 'p-retry'
+
+const ONE_HOUR = 60 * 60 * 1000
+
+const withPgClient = fn => async ({ createPgClient, ...args }) => {
+  const pgClient = await createPgClient()
+  try {
+    return await fn({ pgClient, ...args })
+  } finally {
+    await pgClient.end()
+  }
+}
+
+export const postEvaluate = withPgClient(async ({ pgClient, round, gitCommit }) => {
+  await pgClient.query(`
+    INSERT INTO publish_rsr_rounds
+    (round_index, spark_evaluate_version, measurement_commitments, round_details, providers)
+    VALUES
+    ($1, $2, $3, $4, $5)
+  `, [
+    round.index,
+    gitCommit,
+    round.cids,
+    round.details,
+    round.measurements.reduce((acc, m) => {
+      acc[m.minerId] = acc[m.minerId] || { successful: 0, total: 0 }
+      acc[m.minerId].total++
+      if (m.retrievalResult === 'OK') {
+        acc[m.minerId].successful++
+      }
+      return acc
+    }, {})
```

Comment:
The on-chain data shouldn't be based on committees, is your suggestion still true?

Reply:
Since we want this to be the per-measurement RSR, not per-committee. But I think indeed I'm missing the IPv4 logic, and potentially other parts, as `.total` should only be accepted measurements. I believe we need this:

```js
let total = 0
let successful = 0
for (const m of measurements) {
  if (m.fraudAssessment === 'ok') total++
  if (m.retrievalResult === 'ok') successful++
}
```

Reply:
https://github.com/filecoin-station/spark-evaluate/pull/383/commits/f5e6ee0019f372f2e2fbda5c70ae5f9c0a09e020

Reply:
You are excluding minority results from RSR.

See https://github.com/filecoin-station/spark-evaluate/commit/3b31a57a483fba5d61e1b280b695dcf1ba73d04e

> This commit fixes the problem by changing the implementation of
`updatePublicStats` to iterate over all measurements assigned to all
committees. This way we include all measurements in the calculation:
minority measurements, measurements where no majority was found,
measurements belonging to committees that are too small.

Is it okay to have a different per-measurement RSR in the on-chain commitment than the per-measurement RSR reported by Spark Stats API? If not, then I urge you to share the code building this metric.

Reply:
Let me clarify my previous comment.

In this proposal:

```js
let total = 0
let successful = 0
for (const m of measurements) {
  if (m.fraudAssessment === 'ok') total++
  if (m.retrievalResult === 'ok') successful++
}
```

- `total` counts only accepted measurements in the majority
- `retrievalResult` counts all measurements reporting successfull retrieval, regardless of whether they were accepted/rejected and in minority/majority

The source of the confusion is that `fraudAssessment` combines two steps - 1) whether the measurement is accepted based on tasking, IPv4 subnets, etc. and 2) whether the measurement is in the majority. This is good for deciding which measurements to reward - only those with `ok` assessment, but it makes RSR calculation more complex.

I think it would be worth taking a closer look at this and seeing how to refactor the code to make these two evaluation step outcomes more explicit & distinguished. But that's out of the scope of this PR.